### PR TITLE
feat(js) Add `ArrayBuffer::new`, `::is_view` & `::slice` bindings.

### PR DESCRIPTION
**Pull request by Hywan**
Cf. https://github.com/rustwasm/wasm-bindgen/issues/275

My wish was to include `get ArrayBuffer[@@species]` but I've no idea how to tackle this binding :-). Any recommendation?

### CODE REVIEW

File: tests/all/js_globals/ArrayBuffer.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        .file("src/lib.rs", r#"
+            #![feature(proc_macro, wasm_custom_section)]
+
+            extern crate wasm_bindgen;
+            use wasm_bindgen::prelude::*;
+            use wasm_bindgen::js::ArrayBuffer;
+
+            #[wasm_bindgen]
+            pub fn slice_at_2(arraybuffer: ArrayBuffer) -> ArrayBuffer {
+                ArrayBuffer::slice(&arraybuffer, 2f64)
```
Comment (belfz) [inline tests/all/js_globals/ArrayBuffer.rs]: `slice` is an instance method. Could you please us `arraybuffer.slice(2f64)` instead?

File: tests/all/js_globals/ArrayBuffer.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    project()
+        .file("src/lib.rs", r#"
+            #![feature(proc_macro, wasm_custom_section)]
+
+            extern crate wasm_bindgen;
+            use wasm_bindgen::prelude::*;
+            use wasm_bindgen::js::ArrayBuffer;
+
+            #[wasm_bindgen]
+            pub fn slice_at_2(arraybuffer: ArrayBuffer) -> ArrayBuffer {
```
Comment (belfz) [inline tests/all/js_globals/ArrayBuffer.rs]: this argument should be passed as a reference (`&ArrayBuffer`) :)

File: src/js.rs [hunk 3]
```diff
@@ -254,6 +254,39 @@ extern "C" {
...
+    #[wasm_bindgen(static_method_of = ArrayBuffer, js_name = isView)]
+    pub fn is_view(value: JsValue) -> bool;
+
+    /// The `slice()` method returns a new `ArrayBuffer` whose contents
+    /// are a copy of this `ArrayBuffer`'s bytes from begin, inclusive,
+    /// up to end, exclusive.
+    ///
+    /// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/slice
+    #[wasm_bindgen(method)]
+    pub fn slice(this: &ArrayBuffer, begin: f64) -> ArrayBuffer;
```
Comment (belfz) [inline src/js.rs]: we might need another method here, that takes into account an optional `end` argument - eg. `pub fn slice_with_end(this: &ArrayBuffer, begin: f64, end: f64) -> ArrayBuffer`

Comment (alexcrichton) [reply src/js.rs]: I think the arguments here should be integers as well since they're indices as well, right?

File: src/js.rs [hunk 4]
```diff
@@ -254,6 +254,39 @@ extern "C" {
...
+    ///
+    /// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer
+    #[wasm_bindgen(constructor)]
+    pub fn new(length: f64) -> ArrayBuffer;
+
+    /// The `slice()` method returns a new `ArrayBuffer` whose contents
+    /// are a copy of this `ArrayBuffer`'s bytes from begin, inclusive,
+    /// up to end, exclusive.
+    ///
+    /// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/slice
```
Comment (alexcrichton) [inline src/js.rs]: This URL I think should be https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/isView?

File: src/js.rs (hunk 3 above)
Comment (belfz) [reply src/js.rs]: True!

File: tests/all/js_globals/ArrayBuffer.rs (hunk 1 above)
Comment (Hywan) [reply tests/all/js_globals/ArrayBuffer.rs]: Fixed :-).

File: tests/all/js_globals/ArrayBuffer.rs (hunk 2 above)
Comment (Hywan) [reply tests/all/js_globals/ArrayBuffer.rs]: Fixed :-).

File: src/js.rs (hunk 4 above)
Comment (Hywan) [reply src/js.rs]: Fixed :-). Good catch!

File: src/js.rs (hunk 3 above)
Comment (Hywan) [reply src/js.rs]: Fixed, and fixed :-).

**Comment by Hywan** [comment]
Thanks for the review! It is a perfect illustration of: “Why you should not write a PR 5mn before a meeting”.

**Pull request merged**